tests/test_bytes_vec_alloc.rs

```diff
@@ -1,61 +1,87 @@
+static LEDGER: Ledger = Ledger::new();
 
-struct Ledger;
+#[repr(C)]
+struct Ledger {
+    alloc_table: [(AtomicPtr<u8>, AtomicUsize); 512],
+}
 
-const USIZE_SIZE: usize = mem::size_of::<usize>();
+impl Ledger {
+    const fn new() -> Self {
+        // equivalent to size of (AtomicPtr<u8>, AtomicUsize), hopefully
+        #[cfg(target_pointer_width = "64")]
+        let tricky_bits = 0u128;
 
-unsafe impl GlobalAlloc for Ledger {
-    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
-        if layout.align() == 1 && layout.size() > 0 {
-            // Allocate extra space to stash a record of
-            // how much space there was.
-            let orig_size = layout.size();
-            let size = orig_size + USIZE_SIZE;
-            let new_layout = match Layout::from_size_align(size, 1) {
-                Ok(layout) => layout,
-                Err(_err) => return ptr::null_mut(),
-            };
-            let ptr = System.alloc(new_layout);
-            if !ptr.is_null() {
-                (ptr as *mut usize).write(orig_size);
-                let ptr = ptr.offset(USIZE_SIZE as isize);
-                ptr
-            } else {
-                ptr
+        #[cfg(target_pointer_width = "32")]
+        let tricky_bits = 0u64;
+
+        let magic_table = [tricky_bits; 512];
+
+        // i know this looks bad but all the good ways to do this are unstable or not yet
+        // supported in const contexts (even though they should be!)
+        let alloc_table = unsafe { mem::transmute(magic_table) };
+
+        Self { alloc_table }
+    }
+
+    /// Iterate over our table until we find an open entry, then insert into said entry
+    fn insert(&self, ptr: *mut u8, size: usize) {
+        for (entry_ptr, entry_size) in self.alloc_table.iter() {
+            // SeqCst is good enough here, we don't care about perf, i just want to be correct!
+            if entry_ptr
+                .compare_exchange(null_mut(), ptr, Ordering::SeqCst, Ordering::SeqCst)
+                .is_ok()
+            {
+                entry_size.store(size, Ordering::Relaxed);
+                break;
             }
-        } else {
-            System.alloc(layout)
         }
     }
 
-    unsafe fn dealloc(&self, ptr: *mut u8, layout: Layout) {
-        if layout.align() == 1 && layout.size() > 0 {
-            let off_ptr = (ptr as *mut usize).offset(-1);
-            let orig_size = off_ptr.read();
-            if orig_size != layout.size() {
-                panic!(
-                    "bad dealloc: alloc size was {}, dealloc size is {}",
-                    orig_size,
-                    layout.size()
-                );
+    fn remove(&self, ptr: *mut u8) -> usize {
+        for (entry_ptr, entry_size) in self.alloc_table.iter() {
+            if entry_ptr
+                .compare_exchange(ptr, null_mut(), Ordering::SeqCst, Ordering::SeqCst)
+                .is_ok()
+            {
+                return entry_size.swap(0, Ordering::Relaxed);
```

Comment:
This `swap` might be problematic. Using `load` should be enough. The entry_size will we overwritten when the next allocation is done

Reply:
How so?

Reply:
I realize it isn't needed because we already zero the ptr, but I don't see how this would cause issues.

Reply:
This is a test so after all going down this level of detail might not make sense, but I think it's totally possible that between the `compare_exchange` and the `swap` another thread allocates. `swap` could then zero the `entry_size` of the next allocation.
IDK just mentioned it because I also noticed the hang with the CI

Reply:
Ah, good catch! The hang was because I wasn't removing it because I wasn't handling realloc cases correctly.

Reply:
I think even a load would still have issues, as we could read the allocation from the other thread. I suppose setting this to the max value would be better.

Reply:
Might wanna make the ledger _slightly_ bigger just in case.

Reply:
> I think even a load would still have issues, as we could read the allocation from the other thread

True

Reply:
Thanks, I think this should be fixed in my latest commit!